pytorch_lightning/core/step_result.py

```diff
@@ -0,0 +1,382 @@
+from typing import Optional, Dict
+from torch import Tensor
+import torch
+
+
+class Result(Dict):
+
+    def __init__(self,
+                 minimize: Optional[Tensor] = None,
+                 early_stop_on: Tensor = None,
+                 checkpoint_on: Tensor = None,
+                 hiddens: Optional[Tensor] = None):
+        super().__init__()
+
+        self.early_stop_on = early_stop_on
+        self.checkpoint_on = checkpoint_on
+
+        # TODO: should hiddens detach?
+        self.hiddens = hiddens
+        self.minimize = minimize
+
+    def __reduce_on_callback(self, callback_name, name, metric, log, pbar, reduce_fx):
+        assert isinstance(metric, torch.Tensor), f'{name} must be a torch.Tensor'
+
+        keys = [f'reduce_{callback_name}']
+        if log:
+            keys.append(f'log_{callback_name}')
+        if pbar:
+            keys.append(f'pbar_{callback_name}')
+
+        for key in keys:
+            if key not in self:
+                self[key] = {}
+
+            if 'log' in key or 'pbar' in key:
+                metric = metric.detach()
+
+            metrics = self[key]
+            metrics[name] = metric
+
+        key = f'reduce_fx_{callback_name}'
+        if key not in self:
+            self[key] = {}
+
+        metrics = self[key]
+        metrics[name] = reduce_fx
+
+    def pass_to_epoch_end(self, name, metric):
+        if 'to_epoch_end' not in self:
+            self['to_epoch_end'] = {}
+
+        metrics = self['to_epoch_end']
```

Comment:
still think that having internal `metrics` and in `progress_bar` just save a list of metrics keys (on-demand you would dictionary as you can simply zip it in the property) which may reduce user confusion that log metric and progress bar is different because we accidentally did not update one of them...